File: packages/Autonomo.CSharp/Autonomo.Tests/AutonomoTests.cs
/**
 * Test harness for Autonomo.CSharp
 * Run: dotnet run --project Autonomo.Tests
 */

using System;
using System.Collections.Generic;
using Autonomo;

class Program
{
    static int passed = 0;
    static int failed = 0;

    static void Test(string name, Action fn)
    {
        try
        {
            fn();
            Console.WriteLine($"✅ {name}");
            passed++;
        }
        catch (Exception e)
        {
            Console.WriteLine($"❌ {name}");
            Console.WriteLine($"   {e.Message}");
            failed++;
        }
    }

    static void Assert(bool condition, string message)
    {
        if (!condition) throw new Exception(message);
    }

    static void AssertEqual<T>(T expected, T actual, string? message = null)
    {
        if (!Equals(expected, actual))
        {
            throw new Exception(message ?? $"Expected {expected}, got {actual}");
        }
    }

    static void Main(string[] args)
    {
        Console.WriteLine("\n🧪 Autonomo.CSharp Test Harness\n");

        // Reset singletons using instance methods
        ElementRegistry.Instance.Clear();
        CustomActionsRegistry.Instance.Clear();

        Test("registry starts empty", () =>
        {
            Assert(ElementRegistry.Instance.List().Count == 0, "Registry should be empty");
        });

        Test("RegisterTapHandler adds element", () =>
        {
            var tapped = false;
            var unregister = Registry.RegisterTapHandler("Test.Button", () => { tapped = true; });

            Assert(ElementRegistry.Instance.Has("Test.Button"), "Element should exist");
            Assert(ElementRegistry.Instance.List().Contains("Test.Button"), "Element should be in list");

            // Invoke handler
            var handler = ElementRegistry.Instance.Get("Test.Button");
            handler?.Handler(null);
            Assert(tapped, "Handler should have been called");

            unregister();
            Assert(!ElementRegistry.Instance.Has("Test.Button"), "Element should be removed");
        });

        Test("RegisterFillHandler works with value", () =>
        {
            var value = "";
            var unregister = Registry.RegisterFillHandler(
                "Test.Input",
                (v) => { value = v ?? ""; },
                getValue: () => value
            );

            var handler = ElementRegistry.Instance.Get("Test.Input");
            handler?.Handler("test value");
            AssertEqual("test value", value);
            AssertEqual("test value", handler?.GetValue?.Invoke());

            unregister();
        });

        Test("custom actions work", () =>
        {
            var unregister = CustomActions.Register("testAction", (value) =>
            {
                if (value == "fail")
                    return ActionResult.Fail("Intentional failure");
                return ActionResult.Ok($"Got: {value}");
            });

            var result = CustomActions.Execute("testAction", "hello");
            Assert(result.Success, "Should succeed");
            Assert(result.Message?.Contains("Got: hello") == true, "Message should contain value");

            result = CustomActions.Execute("testAction", "fail");
            Assert(!result.Success, "Should fail");

            unregister();
        });

        Test("state manager tracks screen", () =>
        {
            State.SetScreen("login");
            AssertEqual("login", State.GetScreen());

            var state = State.GetState();
            AssertEqual("login", state.Screen);
        });

        Test("state manager tracks user", () =>
        {
            var user = new UserContext { Id = "123", Email = "test@example.com", Role = "admin" };
            State.SetUser(user);

            var state = State.GetState();
            AssertEqual("123", state.User?.Id);
            AssertEqual("test@example.com", state.User?.Email);
        });

        Test("state manager tracks errors", () =>
        {
            StateManager.Instance.ClearErrors();
            State.AddError("Test error");

            var state = State.GetState();
            Assert(state.Errors.Contains("Test error"), "Errors should contain test error");
        });

        Test("commands execute press", () =>
        {
            var pressed = false;
            Registry.RegisterTapHandler("Cmd.Button", () => { pressed = true; });

            var result = Commands.Press("Cmd.Button");
            Assert(result.Success, "Command should succeed");
            Assert(pressed, "Button should be pressed");

            ElementRegistry.Instance.Unregister("Cmd.Button");
        });

        Test("commands execute fill", () =>
        {
            var value = "";
            Registry.RegisterFillHandler("Cmd.Input", (v) => { value = v ?? ""; });

            var result = Commands.Fill("Cmd.Input", "hello");
            Assert(result.Success, "Command should succeed");
            AssertEqual("hello", value);

            ElementRegistry.Instance.Unregister("Cmd.Input");
        });

        Test("commands return error for missing element", () =>
        {
            var result = Commands.Press("Nonexistent.Button");
            Assert(!result.Success, "Should fail");
            Assert(result.Error?.ToLower().Contains("not found") == true, "Error should mention not found");
        });

        Test("custom actions with metadata", () =>
        {
            var meta = new CustomActionMeta
            {
                Description = "Greets the user",
                Args = new Dictionary<string, string> { ["name"] = "Name to greet" },
                Example = new Dictionary<string, string> { ["name"] = "World" }
            };

            var unregister = CustomActions.Register("greetAction", (value) =>
            {
                return ActionResult.Ok($"Hello, {value}!");
            }, meta);

            // Verify action works
            var result = CustomActions.Execute("greetAction", "World");
            Assert(result.Success, "Should succeed");
            Assert(result.Message?.Contains("Hello, World!") == true, "Message should contain greeting");

            // Verify GetAll returns rich info
            var allActions = CustomActionsRegistry.Instance.GetAll();
            Assert(allActions.Count >= 1, "Should have at least one action");

            var greetInfo = allActions.Find(a => a.Name == "greetAction");
            Assert(greetInfo != null, "Should find greetAction");
            AssertEqual("Greets the user", greetInfo!.Description);
            AssertEqual("Name to greet", greetInfo.Args?["name"]);
            AssertEqual("World", greetInfo.Example?["name"]);

            unregister();
        });

        Console.WriteLine($"\n📊 Results: {passed} passed, {failed} failed\n");
        Environment.Exit(failed > 0 ? 1 : 0);
    }
}

File: packages/autonomo-swift/Tests/AutonomoTests/AutonomoTests.swift
/**
 * Test harness for autonomo-swift
 * Run: swift test
 */

import XCTest
@testable import Autonomo

final class AutonomoTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        ElementRegistry.shared.clear()
        CustomActionsRegistry.shared.clear()
    }
    
    func testRegistryStartsEmpty() {
        XCTAssertEqual(ElementRegistry.shared.list().count, 0, "Registry should be empty")
    }
    
    func testRegisterTapHandler() {
        var tapped = false
        
        let unregister = registerTapHandler("Test.Button") {
            tapped = true
        }
        
        XCTAssertTrue(ElementRegistry.shared.has("Test.Button"), "Element should exist")
        XCTAssertTrue(ElementRegistry.shared.list().contains("Test.Button"), "Element should be in list")
        
        // Invoke handler
        if let handler = ElementRegistry.shared.get("Test.Button") {
            handler.handler(nil)
        }
        XCTAssertTrue(tapped, "Handler should have been called")
        
        unregister()
        XCTAssertFalse(ElementRegistry.shared.has("Test.Button"), "Element should be removed")
    }
    
    func testRegisterFillHandler() {
        var value = ""
        
        let unregister = registerFillHandler(
            "Test.Input",
            getValue: { value }
        ) { v in
            value = v ?? ""
        }
        
        if let handler = ElementRegistry.shared.get("Test.Input") {
            handler.handler("test value")
            XCTAssertEqual(value, "test value")
            XCTAssertEqual(handler.getValue?(), "test value")
        }
        
        unregister()
    }
    
    func testCustomActions() {
        let unregister = registerCustomAction("testAction") { value in
            if value == "fail" {
                return ActionResult.fail("Intentional failure")
            }
            return ActionResult.ok("Got: \(value ?? "")")
        }
        
        var result = CustomActionsRegistry.shared.execute("testAction", value: "hello")
        XCTAssertTrue(result.success, "Should succeed")
        XCTAssertTrue(result.message?.contains("Got: hello") == true)
        
        result = CustomActionsRegistry.shared.execute("testAction", value: "fail")
        XCTAssertFalse(result.success, "Should fail")
        
        unregister()
    }
    
    func testStateManagerTracksScreen() {
        StateManager.shared.setScreen("login")
        XCTAssertEqual(StateManager.shared.getScreen(), "login")
        
        let state = StateManager.shared.getState()
        XCTAssertEqual(state.screen, "login")
    }
    
    func testStateManagerTracksUser() {
        let user = UserContext(id: "123", email: "test@example.com", role: "admin")
        StateManager.shared.setUser(user)
        
        let state = StateManager.shared.getState()
        XCTAssertEqual(state.user?.id, "123")
        XCTAssertEqual(state.user?.email, "test@example.com")
    }
    
    func testStateManagerTracksErrors() {
        StateManager.shared.clearErrors()
        StateManager.shared.addError("Test error")
        
        let state = StateManager.shared.getState()
        XCTAssertTrue(state.errors.contains("Test error"))
    }
    
    func testCommandsExecutePress() {
        var pressed = false
        registerTapHandler("Cmd.Button") { pressed = true }
        
        let result = executeCommand("press", target: "Cmd.Button")
        XCTAssertTrue(result.success, "Command should succeed")
        XCTAssertTrue(pressed, "Button should be pressed")
        
        ElementRegistry.shared.unregister("Cmd.Button")
    }
    
    func testCommandsExecuteFill() {
        var value = ""
        registerFillHandler("Cmd.Input") { v in value = v ?? "" }
        
        let result = executeCommand("fill", target: "Cmd.Input", value: "hello")
        XCTAssertTrue(result.success, "Command should succeed")
        XCTAssertEqual(value, "hello")
        
        ElementRegistry.shared.unregister("Cmd.Input")
    }
    
    func testCommandsReturnErrorForMissingElement() {
        let result = executeCommand("press", target: "Nonexistent.Button")
        XCTAssertFalse(result.success, "Should fail")
        XCTAssertTrue(result.error?.lowercased().contains("not found") == true)
    }
    
    func testCustomActionsWithMetadata() {
        let meta = CustomActionMeta(
            description: "Greets the user",
            args: ["name": "Name to greet"],
            example: ["name": "World"]
        )
        
        let unregister = registerCustomAction("greetAction", meta: meta) { value in
            return ActionResult.ok("Hello, \(value ?? "stranger")!")
        }
        
        // Verify action works
        let result = CustomActionsRegistry.shared.execute("greetAction", value: "World")
        XCTAssertTrue(result.success)
        XCTAssertTrue(result.message?.contains("Hello, World!") == true)
        
        // Verify getAll returns rich info
        let allActions = CustomActionsRegistry.shared.getAll()
        XCTAssertGreaterThanOrEqual(allActions.count, 1)
        
        let greetInfo = allActions.first { $0.name == "greetAction" }
        XCTAssertNotNil(greetInfo)
        XCTAssertEqual(greetInfo?.description, "Greets the user")
        XCTAssertEqual(greetInfo?.args?["name"], "Name to greet")
        XCTAssertEqual(greetInfo?.example?["name"], "World")
        
        unregister()
    }
}
